azkaban-common/src/main/java/azkaban/project/ProjectManager.java

```diff
@@ -86,12 +98,115 @@ public ProjectManager(ProjectLoader loader, Props props) {
+              lastProjectFileCleanTime = currentTime;
+            }
+
+            if (currentTime - CLEANER_THREAD_WAIT_INTERVAL_MS > lastProjectCleanTime) {
+              purgeRemovedProjects();
+              lastProjectCleanTime = currentTime;
+            }
+
+            logger.info("Cleaner Thread completed. Waiting!!!!!");
+            wait(CLEANER_THREAD_WAIT_INTERVAL_MS);
+          } catch (ProjectManagerException e) {
+            logger.error(this.getName() + "failed due to: " + e.getMessage());
+          } catch (InterruptedException e) {
+            logger.info("Interrupted. Probably to shut down.");
+          }
+        }
+      }
+    }
+
+    private void cleanProjectFilesAndRemoveProject() throws ProjectManagerException {
+      User deleter = new User(DELETER);
+      long currentTime = System.currentTimeMillis();
+
+      Set<Integer> scheduledProjectIds = projectLoader.fetchProjectIdsByEventType(EventType.SCHEDULE);
+      logger.info("Scheduled Projects count: " + scheduledProjectIds.size());
+
+      List<Project> projects = projectsById.values()
+              .stream()
+              .filter(project -> !scheduledProjectIds.contains(project.getId()) &&
+                      currentTime - project.getLastModifiedTimestamp() > projectFilesRetentionMs)
+              .collect(Collectors.toList());
+
+      logger.info("Project files to clean: " + projects.size());
+
+      for(Project project : projects) {
+        logger.debug("Cleaning project files for project: " + project.getId());
+        projectLoader.cleanProjectFiles(project);
+        logger.debug("Removing project: " + project.getId());
+        removeProject(project, deleter);
+      }
+    }
+
+    private void purgeRemovedProjects() throws ProjectManagerException {
+      long currentTime = System.currentTimeMillis();
+
+      List<Project> inactiveProjects = projectLoader.fetchAllInactiveProjects();
+      logger.info("Inactive projects count: " + inactiveProjects.size());
+
+      Set<Integer> purgedProjectIds = projectLoader.fetchProjectIdsByEventType(EventType.PURGE);
+      logger.info("Purged projects count: " + purgedProjectIds.size());
+
+      List<Project> projectsToPurge = inactiveProjects
+              .stream()
+              .filter(project -> !purgedProjectIds.contains(project.getId()) &&
+                      currentTime - project.getLastModifiedTimestamp() > removedProjectRetentionMs)
+              .collect(Collectors.toList());
+
+      logger.info("Projects to purge count: " + projectsToPurge.size());
+
+      User deleter = new User(DELETER);
+      for(Project project: projectsToPurge) {
+        logger.debug("Purging project: " + project.getId());
+        purgeProject(project, deleter);
+      }
+    }
   }
 
   private void loadAllProjects() {
     List<Project> projects;
     try {
       projects = projectLoader.fetchAllActiveProjects();
+      logger.info("Active projects count: " + projects.size());
```

Comment:
Do we need this? We have enough logging above

Reply:
This we just added because when Azkaban Webserver boots up, we see from that strace command right now how many projects are loaded, this will just give total projects so that we need not run that mysql query to check how many projects are there